# Multiple Linear Regression

In [1]:
import pyforest

In [2]:
dataset = pd.read_csv('50_Startups.csv')
dataset

<IPython.core.display.Javascript object>

R&D Spend  Administration  Marketing Spend       State     Profit
0   165349.20       136897.80        471784.10    New York  192261.83
1   162597.70       151377.59        443898.53  California  191792.06
2   153441.51       101145.55        407934.54     Florida  191050.39
3   144372.41       118671.85        383199.62    New York  182901.99
4   142107.34        91391.77        366168.42     Florida  166187.94
5   131876.90        99814.71        362861.36    New York  156991.12
6   134615.46       147198.87        127716.82  California  156122.51
7   130298.13       145530.06        323876.68     Florida  155752.60
8   120542.52       148718.95        311613.29    New York  152211.77
9   123334.88       108679.17        304981.62  California  149759.96
10  101913.08       110594.11        229160.95     Florida  146121.95
11  100671.96        91790.61        249744.55  California  144259.40
12   93863.75       127320.38        249839.44     Florida  141585.52
13   91992.39       135495.07        252664.93  California  134307.35
14  119943.24       156547.42        256512.92     Florida  132602.65
15  114523.61       122616.84        261776.23    New York  129917.04
16   78013.11       121597.55        264346.06  California  126992.93
17   94657.16       145077.58        282574.31    New York  125370.37
18   91749.16       114175.79        294919.57     Florida  124266.90
19   86419.70       153514.11             0.00    New York  122776.86
20   76253.86       113867.30        298664.47  California  118474.03
21   78389.47       153773.43        299737.29    New York  111313.02
22   73994.56       122782.75        303319.26     Florida  110352.25
23   67532.53       105751.03        304768.73     Florida  108733.99
24   77044.01        99281.34        140574.81    New York  108552.04
25   64664.71       139553.16        137962.62  California  107404.34
26   75328.87       144135.98        134050.07     Florida  105733.54
27   72107.60       127864.55        353183.81    New York  105008.31
28   66051.52       182645.56        118148.20     Florida  103282.38
29   65605.48       153032.06        107138.38    New York  101004.64
30   61994.48       115641.28         91131.24     Florida   99937.59
31   61136.38       152701.92         88218.23    New York   97483.56
32   63408.86       129219.61         46085.25  California   97427.84
33   55493.95       103057.49        214634.81     Florida   96778.92
34   46426.07       157693.92        210797.67  California   96712.80
35   46014.02        85047.44        205517.64    New York   96479.51
36   28663.76       127056.21        201126.82     Florida   90708.19
37   44069.95        51283.14        197029.42  California   89949.14
38   20229.59        65947.93        185265.10    New York   81229.06
39   38558.51        82982.09        174999.30  California   81005.76
40   28754.33       118546.05        172795.67  California   78239.91
41   27892.92        84710.77        164470.71     Florida   77798.83
42   23640.93        96189.63        148001.11  California   71498.49
43   15505.73       127382.30         35534.17    New York   69758.98
44   22177.74       154806.14         28334.72  California   65200.33
45    1000.23       124153.04          1903.93    New York   64926.08
46    1315.46       115816.21        297114.46     Florida   49490.75
47       0.00       135426.92             0.00  California   42559.73
48     542.05        51743.15             0.00    New York   35673.41
49       0.00       116983.80         45173.06  California   14681.40

In [3]:
# dividing into dependent and independent features

X = dataset.iloc[:,:-1]
y = dataset.iloc[:, 4]

In [4]:
#dividing into one hot encoding or the dummy variables

states = pd.get_dummies(X['State'],drop_first = True)  #to avoid dummy trap we will drop the first column it means that it if two column i.e california and florida have 0 value in one hot encoding then it means we are talking about the newyork so to avoid this we will seperate the newyork from the dataframe
states


<IPython.core.display.Javascript object>

Florida  New York
0         0         1
1         0         0
2         1         0
3         0         1
4         1         0
5         0         1
6         0         0
7         1         0
8         0         1
9         0         0
10        1         0
11        0         0
12        1         0
13        0         0
14        1         0
15        0         1
16        0         0
17        0         1
18        1         0
19        0         1
20        0         0
21        0         1
22        1         0
23        1         0
24        0         1
25        0         0
26        1         0
27        0         1
28        1         0
29        0         1
30        1         0
31        0         1
32        0         0
33        1         0
34        0         0
35        0         1
36        1         0
37        0         0
38        0         1
39        0         0
40        0         0
41        1         0
42        0         0
43        0         1
44        0         0
45        0         1
46        1         0
47        0         0
48        0         1
49        0         0

In [6]:
#dropping the state column 

X = X.drop('State',axis=1)

KeyError: "['State'] not found in axis"

In [7]:
#concatenating all columns with 5 independent features

X = pd.concat([X,states],axis=1)
X

<IPython.core.display.Javascript object>

R&D Spend  Administration  Marketing Spend  Florida  New York
0   165349.20       136897.80        471784.10        0         1
1   162597.70       151377.59        443898.53        0         0
2   153441.51       101145.55        407934.54        1         0
3   144372.41       118671.85        383199.62        0         1
4   142107.34        91391.77        366168.42        1         0
5   131876.90        99814.71        362861.36        0         1
6   134615.46       147198.87        127716.82        0         0
7   130298.13       145530.06        323876.68        1         0
8   120542.52       148718.95        311613.29        0         1
9   123334.88       108679.17        304981.62        0         0
10  101913.08       110594.11        229160.95        1         0
11  100671.96        91790.61        249744.55        0         0
12   93863.75       127320.38        249839.44        1         0
13   91992.39       135495.07        252664.93        0         0
14  119943.24       156547.42        256512.92        1         0
15  114523.61       122616.84        261776.23        0         1
16   78013.11       121597.55        264346.06        0         0
17   94657.16       145077.58        282574.31        0         1
18   91749.16       114175.79        294919.57        1         0
19   86419.70       153514.11             0.00        0         1
20   76253.86       113867.30        298664.47        0         0
21   78389.47       153773.43        299737.29        0         1
22   73994.56       122782.75        303319.26        1         0
23   67532.53       105751.03        304768.73        1         0
24   77044.01        99281.34        140574.81        0         1
25   64664.71       139553.16        137962.62        0         0
26   75328.87       144135.98        134050.07        1         0
27   72107.60       127864.55        353183.81        0         1
28   66051.52       182645.56        118148.20        1         0
29   65605.48       153032.06        107138.38        0         1
30   61994.48       115641.28         91131.24        1         0
31   61136.38       152701.92         88218.23        0         1
32   63408.86       129219.61         46085.25        0         0
33   55493.95       103057.49        214634.81        1         0
34   46426.07       157693.92        210797.67        0         0
35   46014.02        85047.44        205517.64        0         1
36   28663.76       127056.21        201126.82        1         0
37   44069.95        51283.14        197029.42        0         0
38   20229.59        65947.93        185265.10        0         1
39   38558.51        82982.09        174999.30        0         0
40   28754.33       118546.05        172795.67        0         0
41   27892.92        84710.77        164470.71        1         0
42   23640.93        96189.63        148001.11        0         0
43   15505.73       127382.30         35534.17        0         1
44   22177.74       154806.14         28334.72        0         0
45    1000.23       124153.04          1903.93        0         1
46    1315.46       115816.21        297114.46        1         0
47       0.00       135426.92             0.00        0         0
48     542.05        51743.15             0.00        0         1
49       0.00       116983.80         45173.06        0         0

In [8]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3)

<IPython.core.display.Javascript object>

In [11]:
#Multiple linear regression to the training dataset

linear = LinearRegression()
linear.fit(X_train, y_train)

print(y_test)

<IPython.core.display.Javascript object>

14    132602.65
37     89949.14
8     152211.77
9     149759.96
19    122776.86
20    118474.03
21    111313.02
44     65200.33
46     49490.75
5     156991.12
24    108552.04
28    103282.38
2     191050.39
40     78239.91
36     90708.19
Name: Profit, dtype: float64


In [12]:
#predicting the test set results

y_pred = linear.predict(X_test)
y_pred

array([149961.09267936,  89105.21854358, 153358.80826386, 155501.2801956 ,
       115602.37681773, 117075.86708962, 118746.28786726,  63715.99931776,
        56124.90249944, 165240.36588717, 113711.01420243, 101319.68957665,
       183148.67955327,  74473.90456978,  74928.95001   ])

In [17]:
df = pd.DataFrame(y_pred, columns = ['y_pred'])
df

<IPython.core.display.Javascript object>

y_pred
0   149961.092679
1    89105.218544
2   153358.808264
3   155501.280196
4   115602.376818
5   117075.867090
6   118746.287867
7    63715.999318
8    56124.902499
9   165240.365887
10  113711.014202
11  101319.689577
12  183148.679553
13   74473.904570
14   74928.950010

Comparing y_test and y_pred values in Rsquared. (SSmean > SSresidual). R^2 values lie greater than 0.8 if your model is very good

In [19]:
from sklearn.metrics import r2_score
score = r2_score(y_test,y_pred)
score

0.9543567482377536

# Multicollinearity in Linear Regression

In [3]:
df = pd.read_csv("Advertising.csv",index_col = 0)
df

<IPython.core.display.Javascript object>

TV  Radio  Newspaper  Sales
1    230.1   37.8       69.2   22.1
2     44.5   39.3       45.1   10.4
3     17.2   45.9       69.3    9.3
4    151.5   41.3       58.5   18.5
5    180.8   10.8       58.4   12.9
..     ...    ...        ...    ...
196   38.2    3.7       13.8    7.6
197   94.2    4.9        8.1    9.7
198  177.0    9.3        6.4   12.8
199  283.6   42.0       66.2   25.5
200  232.1    8.6        8.7   13.4

[200 rows x 4 columns]

In [4]:
X = df.iloc[:,:-1]
y = df.iloc[:,-1].to_frame()  # to convert series to dataframe

y

Sales
1     22.1
2     10.4
3      9.3
4     18.5
5     12.9
..     ...
196    7.6
197    9.7
198   12.8
199   25.5
200   13.4

[200 rows x 1 columns]

In [5]:
import statsmodels.api as sm

#fit OLS model on X
X = sm.add_constant(X)  #adds const to all rows
X

const     TV  Radio  Newspaper
1      1.0  230.1   37.8       69.2
2      1.0   44.5   39.3       45.1
3      1.0   17.2   45.9       69.3
4      1.0  151.5   41.3       58.5
5      1.0  180.8   10.8       58.4
..     ...    ...    ...        ...
196    1.0   38.2    3.7       13.8
197    1.0   94.2    4.9        8.1
198    1.0  177.0    9.3        6.4
199    1.0  283.6   42.0       66.2
200    1.0  232.1    8.6        8.7

[200 rows x 4 columns]

In [6]:
model = sm.OLS(y, X).fit() #endog is output and exog is input value

In [7]:
model.summary() #const ic the c value

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                  Sales   R-squared:                       0.897
Model:                            OLS   Adj. R-squared:                  0.896
Method:                 Least Squares   F-statistic:                     570.3
Date:                Sun, 20 Aug 2023   Prob (F-statistic):           1.58e-96
Time:                        16:02:38   Log-Likelihood:                -386.18
No. Observations:                 200   AIC:                             780.4
Df Residuals:                     196   BIC:                             793.6
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.9389      0.312      9.422      0.000       2.324       3.554
TV             0.0458      0.001     32.809      0.000       0.043       0.049
Radio          0.1885      0.009     21.893      0.000       0.172       0.206
Newspaper     -0.0010      0.006     -0.177      0.860      -0.013       0.011
==============================================================================
Omnibus:                       60.414   Durbin-Watson:                   2.084
Prob(Omnibus):                  0.000   Jarque-Bera (JB):              151.241
Skew:                          -1.327   Prob(JB):                     1.44e-33
Kurtosis:                       6.332   Cond. No.                         454.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [8]:
#finding correlation between independent features
X.iloc[:,1:].corr()

TV     Radio  Newspaper
TV         1.000000  0.054809   0.056648
Radio      0.054809  1.000000   0.354104
Newspaper  0.056648  0.354104   1.000000

correlation is less than 0.5 in this table it means that there is no multi-collinearity between your independent features hence we can use this model further

## Another example for multi-collinearity

In [9]:
df = pd.read_csv('Salary_Data.csv')
df

<IPython.core.display.Javascript object>

YearsExperience   Age  Salary
0               1.1  21.0   39343
1               1.3  21.5   46205
2               1.5  21.7   37731
3               2.0  22.0   43525
4               2.2  22.2   39891
5               2.9  23.0   56642
6               3.0  23.0   60150
7               3.2  23.3   54445
8               3.2  23.3   64445
9               3.7  23.6   57189
10              3.9  23.9   63218
11              4.0  24.0   55794
12              4.0  24.0   56957
13              4.1  24.0   57081
14              4.5  25.0   61111
15              4.9  25.0   67938
16              5.1  26.0   66029
17              5.3  27.0   83088
18              5.9  28.0   81363
19              6.0  29.0   93940
20              6.8  30.0   91738
21              7.1  30.0   98273
22              7.9  31.0  101302
23              8.2  32.0  113812
24              8.7  33.0  109431
25              9.0  34.0  105582
26              9.5  35.0  116969
27              9.6  36.0  112635
28             10.3  37.0  122391
29             10.5  38.0  121872

In [12]:
X = df.iloc[:,:-1]
y = df.iloc[:,-1].to_frame()

y

Salary
0    39343
1    46205
2    37731
3    43525
4    39891
5    56642
6    60150
7    54445
8    64445
9    57189
10   63218
11   55794
12   56957
13   57081
14   61111
15   67938
16   66029
17   83088
18   81363
19   93940
20   91738
21   98273
22  101302
23  113812
24  109431
25  105582
26  116969
27  112635
28  122391
29  121872

In [14]:
X = sm.add_constant(X)
model = sm.OLS(y, X).fit()
model

In [15]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                 Salary   R-squared:                       0.960
Model:                            OLS   Adj. R-squared:                  0.957
Method:                 Least Squares   F-statistic:                     323.9
Date:                Sun, 20 Aug 2023   Prob (F-statistic):           1.35e-19
Time:                        16:13:47   Log-Likelihood:                -300.35
No. Observations:                  30   AIC:                             606.7
Df Residuals:                      27   BIC:                             610.9
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
===================================================================================
                      coef    std err          t      P>|t|      [0.025      0.975]
-----------------------------------------------------------------------------------
const           -6661.9872   2.28e+04     -0.292      0.773   -5.35e+04    4.02e+04
YearsExperience  6153.3533   2337.092      2.633      0.014    1358.037    1.09e+04
Age              1836.0136   1285.034      1.429      0.165    -800.659    4472.686
==============================================================================
Omnibus:                        2.695   Durbin-Watson:                   1.711
Prob(Omnibus):                  0.260   Jarque-Bera (JB):                1.975
Skew:                           0.456   Prob(JB):                        0.372
Kurtosis:                       2.135   Cond. No.                         626.
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

if we have a high number of std error then there's a huge multicollinearity

In [16]:
X.iloc[:,1:].corr()

YearsExperience       Age
YearsExperience         1.000000  0.987258
Age                     0.987258  1.000000

we have more correlation between both the features and one feature needs to be dropped because the correlation value is becoming same. so there are two remedies:

1. you dont do anything
2. compare the p-value of both the features if one feature has higher p-value then drop that feature so in this case we have age as a higher p-value so we will drop beacause the years of experience is covering 98% of the properties of age so we will train the model based on that.